File: Farming.Domain/Entities/Fertilizer.cs
﻿using Farming.Domain.ValueObjects.Fertilizer;
using Farming.Shared.Abstractions.Domain;

namespace Farming.Domain.Entities
{
    public class Fertilizer : AggregateRoot<FertilizerId>
    {
        public FertilizerId Id { get; }
        public FertilizerTypeId FertilizerTypeId { get; }
        public FertilizerRequiredAmountPerHectare RequiredAmountPerHectare { get; }
        public FertilizerName Name { get; }
        public FertilizerDescription Description { get; }

        public FertilizerType FertilizerType { get; }
        public ICollection<Plant> SuitablePlants { get; }
        public ICollection<FertilizerWarehouseDelivery> FertilizerWarehouseDeliveries { get; }
        public ICollection<FertilizerWarehouseState> FertilizerWarehouseStates { get; }
        public ICollection<FertilizerAction> FertilizerActions { get; }

        public Fertilizer(FertilizerRequiredAmountPerHectare requiredAmountPerHectare, FertilizerName name, FertilizerDescription description)
        {
            Id = new FertilizerId(Guid.NewGuid());
            RequiredAmountPerHectare = requiredAmountPerHectare;
            Name = name;
            Description = description;

            SuitablePlants = new HashSet<Plant>();
            FertilizerWarehouseDeliveries = new HashSet<FertilizerWarehouseDelivery>();
            FertilizerWarehouseStates = new HashSet<FertilizerWarehouseState>();
            FertilizerActions = new HashSet<FertilizerAction>();
        }

        public void AddSutiablePlants(IEnumerable<Plant> plants)
        {
            foreach (var plant in plants)
            {
                SuitablePlants.Add(plant);
            }
        }
    }
}

File: Farming.Frontend/src/app/core/models/fertilizer.ts
export interface FertilizerDto {
  id: string;
  name: string;
  requiredAmountPerHectare: number;
  description: string;
}

export interface FertilizerStateDto {
  fertilizerId: string;
  fertilizerTypeId: string;
  fertilizerName: string;
  fertilizerTypeName: string;
  fertilizerDescription: string;
  fertilizerTypeDescription: string;
  quantity: number;
  requiredAmountPerHectare: number;
  enoughForArea: number;
}

export interface FertilizerActionDto {
  landId: string;
  fertilizerId: string;
  fertilizerWarehouseId: string;
  quantity: number;
}
